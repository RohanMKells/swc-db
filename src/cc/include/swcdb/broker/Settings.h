/*
 * SWC-DBÂ© Copyright since 2019 Alex Kashirin <kashirin.alex@gmail.com>
 * License details at <https://github.com/kashirin-alex/swc-db/#license>
 */

#ifndef swcdb_broker_Settings_h
#define swcdb_broker_Settings_h

#include "swcdb/core/config/Settings.h"
#include "swcdb/core/comm/Settings.h"

#include "swcdb/fs/Settings.h"
#include "swcdb/db/client/Settings.h"
#include "swcdb/db/Types/Encoder.h"


namespace SWC{ namespace Config {


void Settings::init_app_options(){
  init_comm_options();
  init_client_options();

  file_desc.add_options()
    ("swc.bkr.cfg", str(), "Specific cfg-file for Broker")
    ("swc.bkr.cfg.dyn", strs(), "Specific dyn. cfg-file for Broker")

    ("swc.bkr.reactors", i32(8), "Number of Communication Reactors")
    ("swc.bkr.workers", i32(32), "Number of Workers a Reactor")
    ("swc.bkr.handlers", i32(32), "Number of App Handlers")
    ("swc.bkr.clients.handlers", i32(32), "Number of DB-Client Handlers")

    ("swc.bkr.comm.encoder",
      g_enum(
        int(SWC_DEFAULT_COMM_ENCODER),
        nullptr,
        Core::Encoder::from_string_encoding,
        Core::Encoder::repr_encoding),
     "The encoding to use in communication, options PLAIN/ZSTD/SNAPPY/ZLIB")

    ("swc.bkr.ram.allowed.percent", g_i32(33),
     "Memory RSS % allowed without freeing/releasing")
    ("swc.bkr.ram.reserved.percent", g_i32(33),
     "Memory Total % reserved, threshold of low-memory enter state")
    ("swc.bkr.ram.release.rate", g_i32(100),
     "Memory release-rate (malloc dependable)")

    ("swc.bkr.metrics.enabled", boo(true),
     "Enable or Disable Metrics Monitoring")

    ("swc.bkr.metrics.report.interval", g_i32(300),
     "Metrics Reporting Interval in Seconds")
  ;

}

void Settings::init_post_cmd_args() { }

}}

#endif // swcdb_broker_Settings_h
