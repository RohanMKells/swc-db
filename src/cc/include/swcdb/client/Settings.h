/*
 * Copyright (C) 2019 SWC-DB (author: Kashirin Alex (kashirin.alex@gmail.com))
 */

#ifndef swc_client_Settings_h
#define swc_client_Settings_h

#include "swcdb/core/config/Settings.h"
namespace SWC{ namespace Config {

void Settings::init_client_options() {
  file_desc.add_options()
    ("swc.mngr.host", g_strs(gStrings()), 
     "Manager Host: \"[cols range]|(hostname or ips-csv)|port\"")
    ("swc.mngr.port", i16(15000), 
     "Manager default port if not defined in swc.mngr.host")
     
    ("swc.client.Rgr.connection.timeout", g_i32(10000), 
     "Ranger client connect timeout")
    ("swc.client.Rgr.connection.probes", g_i32(1), 
     "Ranger client connect probes")
    ("swc.client.Rgr.connection.keepalive", g_i32(30000), 
     "Ranger client connection keepalive for ms since last action")

    ("swc.client.Mngr.connection.timeout", g_i32(10000), 
     "Manager client connect timeout")
    ("swc.client.Mngr.connection.probes", g_i32(1), 
     "Manager client connect probes")
    ("swc.client.Mngr.connection.keepalive", g_i32(30000), 
     "Manager client connection keepalive for ms since last action")
    ("swc.client.schema.expiry", g_i32(1800000), 
     "Schemas expiry in ms")
    ;
} 
}}

#endif // swc_app_manager_Settings_h