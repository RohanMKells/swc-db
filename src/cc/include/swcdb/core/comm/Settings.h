/*
 * Copyright (C) 2019 SWC-DB (author: Kashirin Alex (kashirin.alex@gmail.com))
 */

#ifndef swc_core_comm_Settings_h
#define swc_core_comm_Settings_h

#include "swcdb/core/config/Settings.h"

namespace SWC{ namespace Config {

//void Settings::init_comm_options();

}}


#ifdef SWC_IMPL_SOURCE
#include "swcdb/core/comm/Settings.cc"
#endif 

#endif // swc_core_comm_Settings_h