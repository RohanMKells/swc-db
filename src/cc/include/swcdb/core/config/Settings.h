/*
 * Copyright Since 2019 SWC-DBÂ© [author: Kashirin Alex kashirin.alex@gmail.com]
 * License details at <https://github.com/kashirin-alex/swc-db/#license>
 */

#ifndef swc_core_config_Config_h
#define swc_core_config_Config_h

#include <memory>
#include "swcdb/core/Error.h"

#include "swcdb/core/config/Properties.h"
#include "swcdb/core/config/PropertiesParser.h"


namespace SWC { namespace Config {

class Settings final : public Properties {

  public:

  ParserConfig    cmdline_desc;
  ParserConfig    file_desc;

  std::string     install_path;
  std::string     executable;

  Settings();

  ~Settings();

  void init(int argc, char *argv[]);

  void init_options();

  void init_app_options();

  void init_comm_options();
  
  void init_fs_options();

  void init_client_options();
  
  void init_post_cmd_args();
 
  void parse_args(int argc, char *argv[]);

  void load_files_by(const std::string& fileprop, 
                     bool allow_unregistered);

  void parse_file(const std::string& fname, const std::string& onchg);

  void init_process();

  std::string usage_str(const char *usage = 0);

  void check_dynamic_files();

  private:

  Parser::Options      m_cmd_args;

  struct DynFile {
    const std::string filename;
    time_t            modified;
    bool operator==(const DynFile& other) const;
    bool operator==(const std::string& other) const;
  };
  std::vector<DynFile> m_dyn_files;
};


}


namespace Env {

class Config final {
  
  public:

  typedef std::shared_ptr<Config> Ptr;

  static void init(int argc, char** argv) {
    get()->m_settings->init(argc, argv);
  }

  static void set(Ptr env){
    m_env = env;
  }

  static Ptr get(){
    if(m_env == nullptr)
      m_env = std::make_shared<Config>();
    return m_env;
  }

  static SWC::Config::Settings* settings() {
    SWC_ASSERT(m_env != nullptr);
    return m_env->m_settings;
  }

  Config() : m_settings(new SWC::Config::Settings()){}

  ~Config() {
    if(m_settings != nullptr)
      delete m_settings;
  }

  private:
  SWC::Config::Settings*  m_settings = nullptr;
  inline static Ptr       m_env = nullptr;

};
}

}

#ifdef SWC_IMPL_SOURCE
#include "swcdb/core/config/Settings.cc"
#endif 

#endif // swc_core_config_Config_h
