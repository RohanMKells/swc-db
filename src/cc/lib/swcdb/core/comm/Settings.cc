/*
 * Copyright Since 2019 SWC-DBÂ© [author: Kashirin Alex kashirin.alex@gmail.com]
 */


#include "swcdb/core/config/Settings.h"

namespace SWC{ namespace Config {

void Settings::init_comm_options() {
  file_desc.add_options()
    ("addr", strs(), 
     "IP-port, addr to listen on else resolved(hostname):swc.ServiceName.port")
    ("host", str(),  
     "host:port to listen on IPv4+6, swc.ServiceName.port applied if port not specified")
    
    ("swc.comm.ssl", boo(false), "Use SSL in comm layer")
    ("swc.comm.ssl.secure.network", strs({"127.0.0.0/8"}),
     "Networks that do not require SSL")
    ("swc.comm.ssl.ciphers", str(), "Ciphers to use")
    ("swc.comm.ssl.ca", str(), "CA, used if set")
    ("swc.comm.ssl.pem", str("server.pem"), "Servers PEM file")
    ("swc.comm.ssl.subject_name", str(), 
      "CRT/Cluster's domain-name, if set SRV-CRT is verified")
     
  ;
}


}}