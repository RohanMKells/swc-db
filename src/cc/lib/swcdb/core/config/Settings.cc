/*
 * Copyright (C) 2019 SWC-DB (author: Kashirin Alex (kashirin.alex@gmail.com))
 */


#include "swcdb/core/config/Settings.h"

#include "swcdb/core/FileUtils.h"


namespace SWC { namespace Config {

Settings::Settings() { 
  cmdline_desc.definition(usage_str());
}

Settings::~Settings() { }

void Settings::init(int argc, char *argv[]) {
  
  char path[1024];
  errno = 0;
  ssize_t r = readlink("/proc/self/exe", path, 1024);
  if(r == -1 || errno)
    throw;

  install_path = std::string(path, r); // install_path = std::string(argv[0]);
 
  auto at = install_path.find_last_of("/");
  if(at == std::string::npos) {
    executable = install_path;
    install_path = ".";
  } else {
    executable = install_path.substr(at?at+1:at, install_path.length());
    install_path = install_path.substr(0, at);
    at = install_path.find_last_of("/");
    install_path = at == std::string::npos ? "." : install_path.substr(0, at);
  }

  init_options();

  init_app_options();

  parse_args(argc, argv);

  init_post_cmd_args();

  bool verbose = properties.get<gBool>("verbose");
  if(verbose && properties.get_bool("quiet")) {
    verbose = false;
    properties.set("verbose", (gBool)false);
  }

  gEnumExtPtr loglevel = properties.get_ptr<gEnumExt>("logging-level");
  if(properties.get_bool("debug"))
    loglevel->set_value(LOG_DEBUG);

  if(loglevel->get() == -1){
    SWC_LOG_OUT(LOG_ERROR) << "unknown logging level: "<< loglevel->str() << SWC_LOG_OUT_END;
    std::quick_exit(EXIT_SUCCESS);
  }

  Logger::logger.set_level(loglevel->get());
  loglevel->set_cb_on_chg([](int value){Logger::logger.set_level(value);});
}

void Settings::init_options() {
  gEnumExt logging_level(LOG_INFO);
  logging_level.set_from_string(Logger::logger.from_string)
               .set_repr(Logger::logger.repr);

  cmdline_desc.add_options()
    ("help,h", "Show this help message and exit")
    ("help-config", "Show help message for config properties")
    ("version,v", "Show version information and exit")

    ("daemon ", boo()->zero_token(), "Start process in background mode")
    ("verbose", g_boo(false)->zero_token(), "Show more verbose output")
    ("debug", boo(false)->zero_token(), "Shortcut to --logging-level debug")
    ("quiet", boo(false)->zero_token(), "Negate verbose")

    ("logging-level,l", g_enum_ext(logging_level), 
     "Logging level: debug|info|notice|warn|error|crit|alert|fatal")
    
    ("swc.cfg.path", str(install_path+"/etc/swcdb/"), "Path of configuration files")
    ("swc.cfg", str("swc.cfg"), "Main configuration file")

    ("swc.logging.path", str(install_path + "/var/log/swcdb/"), "Path of log files")

    //("induce-failure", str(), "Arguments for inducing failure")
    /* Interactive-Shell options
    ("silent", boo()->zero_token(),
     "as Not Interactive or Show as little output as possible") 
    */
    ;
    
  alias("logging-level", "swc.logging.level");

  file_desc.add_options()
    ("swc.logging.level", g_enum_ext(logging_level), 
     "Logging level: debug|info|notice|warn|error|crit|alert|fatal");
}

void Settings::parse_args(int argc, char *argv[]) {

  Parser prs(false);
  prs.config.add(cmdline_desc);
  prs.config.add(file_desc);
  prs.parse_cmdline(argc, argv);
 
  prs.own_options(m_cmd_args);
  
  properties.load_from(prs.get_options());
    
  // some built-in behavior
  if (has("help")) {
    std::cout << cmdline_desc << std::flush;
    std::quick_exit(EXIT_SUCCESS);
  }

  if (has("help-config")) {
    std::cout << file_desc << std::flush;
    std::quick_exit(EXIT_SUCCESS);
  }

  if (has("version")) {
    std::cout << "Version: " << SWC::VERSION << std::endl;
    std::quick_exit(EXIT_SUCCESS);
  }

  if(has("swc.cfg")) 
    parse_file(properties.get_str("swc.cfg"), "swc.OnFileChange.cfg");
}

void Settings::parse_file(const std::string &name, const std::string &onchg) {
  if(name.empty())
    return;

  std::string fname;
  if(name.front() != '/') 
    fname.append(properties.get_str("swc.cfg.path"));
  fname.append(name);
  
  if(!FileUtils::exists(fname))
    SWC_THROWF(Error::FS_FILE_NOT_FOUND, 
              "cfg file=%s not found", fname.c_str());
    
  properties.load(fname, file_desc, cmdline_desc, false);
  if(!onchg.empty())
    load_files_by(onchg, false);
    
  properties.load_from(m_cmd_args);  // Inforce cmdline properties 
}

void Settings::load_files_by(const std::string &fileprop,  
                             bool allow_unregistered) {
    if(fileprop.empty() || !has(fileprop)) 
      return;

    std::string fname;
    Strings files = properties.get_strs(fileprop);
    for (auto it=files.begin(); it<files.end(); it++) {
      fname.clear();
      if(it->front() != '/') 
        fname.append(properties.get_str("swc.cfg.path"));
      fname.append(*it);

	    try {
        properties.load(fname, file_desc, cmdline_desc, allow_unregistered);
      } catch (std::exception &e) {
		    SWC_LOGF(LOG_WARN, "%s has bad cfg file %s: %s", 
                  fileprop.c_str(), it->c_str(), e.what());
      }
    }
  }

void Settings::init_process() {
  bool daemon = properties.has("daemon");

  if(daemon) {
    if(pid_t p = fork())
      exit(0);
  }
  auto pid = getpid();

  executable.append(".");
  executable.append(std::to_string((size_t)pid));

  Logger::logger.initialize(executable);
  if(daemon)
    Logger::logger.daemon(properties.get_str("swc.logging.path"));
    
  if(properties.get<gBool>("verbose")) {
    SWC_LOG_OUT(LOG_NOTICE) 
      << "Initialized " << executable << " (" << SWC::VERSION << ")\n"
      << "Process Settings: \n" << properties.to_string() << SWC_LOG_OUT_END;
  }
}

const bool Settings::has(const std::string &name) const {
  return properties.has(name);
}

const bool Settings::defaulted(const std::string &name) {
  return properties.defaulted(name);
}

void Settings::alias(const std::string &minor, const std::string &major) {
  properties.alias(minor, major);
} 

std::string Settings::usage_str(const char *usage) {
  if (!usage)
    usage = "Usage: %s [options]\n\nOptions:";

  if (strstr(usage, "%s"))
    return format(usage, executable.c_str());

  return usage;
}

}} // namespace SWC::Config
