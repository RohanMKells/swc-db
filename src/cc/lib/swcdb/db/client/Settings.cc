/*
 * SWC-DBÂ© Copyright since 2019 Alex Kashirin <kashirin.alex@gmail.com>
 * License details at <https://github.com/kashirin-alex/swc-db/#license>
 */

#include "swcdb/core/Encoder.h"
#include "swcdb/db/client/Settings.h"

namespace SWC{ namespace Config {

void Settings::init_client_options() {
  cmdline_desc.add_options()
    ("swc.cfg.dyn.period", g_i32(600000),
     "Dynamic cfg-file check interval in ms, zero without")

    ("swc.mngr.host", g_strs(Strings()),
     "Manager Host: \"[cols range]|(hostname or ips-csv)|port\"")
    ("swc.mngr.port", i16(15000),
     "Manager default port if not defined in swc.mngr.host")
    ("swc.rgr.port", i16(16000), "Ranger port")

    ("swc.bkr.host", g_strs(Strings()),
     "Broker Host: \"(hostname or ips-csv)|port\"")
    ("swc.bkr.port", i16(17000), "Broker port")

    ("swc.client.Bkr.connection.timeout", g_i32(10000),
     "Broker client connect timeout")
    ("swc.client.Bkr.connection.probes", g_i32(1),
     "Broker client connect probes")
    ("swc.client.Bkr.connection.keepalive", g_i32(30000),
     "Broker client connection keepalive for ms since last action")

    ("swc.client.Bkr.comm.encoder",
      g_enum(
        int(SWC_DEFAULT_COMM_ENCODER),
        nullptr,
        Core::Encoder::from_string_encoding,
        Core::Encoder::repr_encoding),
     "The encoding to use in communication, options PLAIN/ZSTD/SNAPPY/ZLIB")

    ("swc.client.Rgr.connection.timeout", g_i32(10000),
     "Ranger client connect timeout")
    ("swc.client.Rgr.connection.probes", g_i32(1),
     "Ranger client connect probes")
    ("swc.client.Rgr.connection.keepalive", g_i32(30000),
     "Ranger client connection keepalive for ms since last action")

    ("swc.client.Rgr.comm.encoder",
      g_enum(
        int(SWC_DEFAULT_COMM_ENCODER),
        nullptr,
        Core::Encoder::from_string_encoding,
        Core::Encoder::repr_encoding),
     "The encoding to use in communication, options PLAIN/ZSTD/SNAPPY/ZLIB")

    ("swc.client.Rgr.range.res.expiry", g_i32(1800000),
     "Range Ranger resolution expiry in ms")

    ("swc.client.Mngr.connection.timeout", g_i32(10000),
     "Manager client connect timeout")
    ("swc.client.Mngr.connection.probes", g_i32(1),
     "Manager client connect probes")
    ("swc.client.Mngr.connection.keepalive", g_i32(30000),
     "Manager client connection keepalive for ms since last action")

    ("swc.client.Mngr.comm.encoder",
      g_enum(
        int(SWC_DEFAULT_COMM_ENCODER),
        nullptr,
        Core::Encoder::from_string_encoding,
        Core::Encoder::repr_encoding),
     "The encoding to use in communication, options PLAIN/ZSTD/SNAPPY/ZLIB")

    ("swc.client.Mngr.range.master.expiry", g_i32(1800000),
     "Cached Master Range expiry in ms")

    ("swc.client.schema.expiry", g_i32(1800000),
     "Schemas expiry in ms")

    ("swc.client.send.buffer", g_i32(8388608),
     "Client send buffer size in bytes")
    ("swc.client.send.ahead", g_i8(3),
     "Client send number of buffers ahead")
    ("swc.client.send.timeout", g_i32(800000),
     "Client send timeout in ms")
    ("swc.client.send.timeout.bytes.ratio", g_i32(1000),
     "Timeout ratio to bytes, bytes/ratio=ms added to send timeout")

    ("swc.client.request.again.delay", g_i32(500),
     "Client request again delay size in ms")

    ("swc.client.recv.buffer", g_i32(8388608),
     "Client receive buffer size in bytes")
    ("swc.client.recv.ahead", g_i8(3),
     "Client receive number of buffers ahead")
    ("swc.client.recv.timeout", g_i32(800000),
     "Client receive timeout in ms")
  ;
}


}} // namespace SWC::Config
