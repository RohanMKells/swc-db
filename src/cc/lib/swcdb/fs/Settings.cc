/*
 * Copyright Since 2019 SWC-DBÂ© [author: Kashirin Alex kashirin.alex@gmail.com]
 */


#include "swcdb/core/config/Settings.h"

namespace SWC{ namespace Config {


void Settings::init_fs_options() {
  file_desc.add_options()
    ("swc.fs.path.data", str("swcdb/"), 
     "SWC-DB data-path, within the FS(specific) base-path")

    ("swc.fs", str(), "main FileSystem: local|hadoop|hadoop_jvm|ceph|broker|custom")

    ("swc.fs.local.cfg", str(), "Specific cfg-file for FS-local")
    ("swc.fs.lib.local", str(), "FS-local Lib-path based on fs/FileSystem.h")

    ("swc.fs.hadoop.cfg", str(), "Specific cfg-file for FS-hadoop")
    ("swc.fs.lib.hadoop", str(), "FS-hadoop Lib-path based on fs/FileSystem.h")

    ("swc.fs.hadoop_jvm.cfg", str(), "Specific cfg-file for FS-hadoop_jvm")
    ("swc.fs.lib.hadoop_jvm", str(), "FS-hadoop-JVM Lib-path based on fs/FileSystem.h")

    ("swc.fs.ceph.cfg", str(), "Specific cfg-file for FS-ceph")
    ("swc.fs.lib.ceph", str(), "FS-ceph Lib-path based on fs/FileSystem.h")

    ("swc.fs.custom.cfg", str(), "Specific cfg-file for FS-custom")
    ("swc.fs.lib.custom", str(), "FS-custom Lib-path based on fs/FileSystem.h")

    ("swc.fs.broker.cfg", str(), "Specific cfg-file for FS-broker")
    ("swc.fs.broker.underlying", str(), 
      "as main FileSystem, without 'broker': local|hadoop|ceph|custom")
    ("swc.fs.lib.broker", str(), "FS-broker Lib-path based on fs/FileSystem.h")
  
  ;
}

}}