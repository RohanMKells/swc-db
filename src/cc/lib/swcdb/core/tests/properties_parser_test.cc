/*
 * Copyright (C) 2019 SWC-DB (author: Kashirin Alex (kashirin.alex@gmail.com))
 */


#include "swcdb/core/config/Settings.h"

namespace SWC{ namespace Config {

void Settings::init_app_options() {
  cmdline_desc
  .definition("Usage: %s [Options] [args]\nOptions")
  .add_options()
   ("i16", i16(1), "16-bit integer")
   ("i32", i32(), "32-bit integer")
   ("i64", i64(1), "64-bit integer")
   ("int64s", i64s(), "a list of 64-bit integers")
   ("boo,b", boo(false), "a boolean arg")
   ("string,str,s", str("A Default std::string Value"), "a string arg")
   ("strs", strs(), "a list of strings")
   ("float64", f64(3.12), "a double arg")
   ("f64s", f64s(), "a list of doubles")
   ("skip1", boo()->zero_token(), "a skipped token")
   ("skip2", "a skipped token")
   ("false", boo(false)->zero_token(), "token defaults to false")
   ("true", boo(true)->zero_token(), "token defaults to true")
   ("is_true", boo(false)->zero_token(), "if set token true")

   ("app",    str(),  "application filename, zero pos skipped")
   ("app",    0)
   ("args",  -1 ) // -1, accumalate unknonw options
   ("args",   strs(), "rest of arguments")
   ("arg-1",  str(),  "argument one")
   ("arg-1",  1)
   ;   
   //("a.cmd.arg.qouted", str(), "a qouted string arg with spaces") require cmake escaping to testdiff  
}

void Settings::init_post_cmd_args(){ }
} }

using namespace SWC;

int main(int argc, char *argv[]) {
  Env::Config::init(argc, argv);
  auto settings = Env::Config::settings();
  
  // remove env-dynamic configs
  settings->cmdline_desc.remove("swc.logging.path");
  settings->cmdline_desc.remove("swc.cfg.path");
  

  std::cout << std::string("\nConfig::cmdline_desc");
  std::cout << settings->cmdline_desc;
  settings->print(std::cout);

  
  // cfg file parse test definitions
  settings->file_desc.add_options()
   ("is_true_bool", boo(false), "a boolean arg")
   ("is_yes_bool", boo(false), "a boolean arg")
   ("is_one_bool", boo(false), "a boolean arg")
   ("is_no_bool", boo(true), "a boolean arg")
   ("is_rest_bool", boo(true), "a boolean arg")

   ("boo", boo(false), "a boolean arg")
   ("i16", i16(1), "16-bit integer")
   ("i32", i32(), "32-bit integer")
   ("i64", i64(1), "64-bit integer")
   ("int64s", i64s(), "a list of 64-bit integers")
   ("string", str("A Default std::string Value"), "a string arg")
   ("string.qouted", str("A Default std::string Value"), "a string arg")
   ("strs", strs(), "a list of strings")
   ("float64", f64(3.12), "a double arg")
   ("f64s", f64s(), "a list of doubles")
   ("missing", str("a missing string in cfg file, default applies"), "a string arg")
   ("aGroupOne.arg.1", i16(0), "a group arg")
   ("aGroupOne.arg.2", i16(0), "a group arg")
   ("aGroupOne.arg.3", i16(0), "a group arg")
   ("aGroupOne.arg.4", i16(0), "a group arg")
   ("aGroupOne.arg.5", i16(0), "a group arg")
    ;
    
  std::cout << std::string("\nConfig::file_desc");
  std::cout << settings->file_desc;

  std::ifstream in("properties_parser_test.cfg");
  Config::Parser prs_file(false);
  prs_file.config.add(settings->file_desc);
  prs_file.config.add(settings->cmdline_desc);
  prs_file.parse_filedata(in);
  
  std::cout << std::string("\nConfig::file_desc");
  std::cout << settings->file_desc; // original configuration (left untouched)

  std::cout << prs_file.config;  // configurations options with values
  std::cout << prs_file; // Raw(std::strings) Parsed
  prs_file.print_options(std::cout);
  
  Properties props;
  props.set("dummy", boo(true));
  std::cout << props.to_string("dummy") << "\n";

  props.load_from(prs_file.get_options());

  props.print(std::cout, true);

  return 0;
}