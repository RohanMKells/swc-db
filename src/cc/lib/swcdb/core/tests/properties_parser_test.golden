
Config::cmdline_descUsage: %s [Options] [args]
Options
  --app                 application filename, zero pos skipped                          
  --arg-1               argument one                                                    
  --args                rest of arguments                                             []
  --boo -[b]            a boolean arg                                                 false
  --daemon              Start process in background mode                              true
  --debug               Shortcut to --logging-level debug                             false
  --f64s                a list of doubles                                             []
  --false               token defaults to false                                       false
  --float64             a double arg                                                  3.12
  --help -[h]           Show this help message and exit                               true
  --help-config         Show help message for config properties                       true
  --i16                 16-bit integer                                                1 
  --i32                 32-bit integer                                                0 
  --i64                 64-bit integer                                                1 
  --int64s              a list of 64-bit integers                                     []
  --is_true             if set token true                                             false
  --logging-level -[l]  Logging level: debug|info|notice|warn|error|crit|alert|fatal  INFO  # (6)
  --quiet               Negate verbose                                                false
  --skip1               a skipped token                                               true
  --skip2               a skipped token                                               true
  --string -[str, s]    a string arg                                                  A Default std::string Value
  --strs                a list of strings                                             []
  --swc.cfg             Main configuration file                                       swc.cfg
  --true                token defaults to true                                        true
  --verbose             Show more verbose output                                      false
  --version -[v]        Show version information and exit                             true
arg-1=argOne
boo=true
f64s=[1.123, 1123, 1.123e+06, 1.123e+09]
float64=1.1234
i16=1000
i32=64000
i64=1000000000
int64s=[1000000000, 2000000000, 3000000000]
is_true=true
string=aStringNoSpaces
strs=[1, 2, 3, 4, 5, 6]
swc.cfg=./properties_parser_test.cfg

Config::file_desc
  --aGroupOne.arg.1    a group arg                                                   0 
  --aGroupOne.arg.2    a group arg                                                   0 
  --aGroupOne.arg.3    a group arg                                                   0 
  --aGroupOne.arg.4    a group arg                                                   0 
  --aGroupOne.arg.5    a group arg                                                   0 
  --boo                a boolean arg                                                 false
  --f64s               a list of doubles                                             []
  --float64            a double arg                                                  3.12
  --i16                16-bit integer                                                1 
  --i32                32-bit integer                                                0 
  --i64                64-bit integer                                                1 
  --int64s             a list of 64-bit integers                                     []
  --is_no_bool         a boolean arg                                                 true
  --is_one_bool        a boolean arg                                                 false
  --is_rest_bool       a boolean arg                                                 true
  --is_true_bool       a boolean arg                                                 false
  --is_yes_bool        a boolean arg                                                 false
  --missing            a string arg                                                  a missing string in cfg file, default applies
  --string             a string arg                                                  A Default std::string Value
  --string.qouted      a string arg                                                  A Default std::string Value
  --strs               a list of strings                                             []
  --swc.logging.level  Logging level: debug|info|notice|warn|error|crit|alert|fatal  INFO  # (6)

Config::file_desc
  --aGroupOne.arg.1    a group arg                                                   0 
  --aGroupOne.arg.2    a group arg                                                   0 
  --aGroupOne.arg.3    a group arg                                                   0 
  --aGroupOne.arg.4    a group arg                                                   0 
  --aGroupOne.arg.5    a group arg                                                   0 
  --boo                a boolean arg                                                 false
  --f64s               a list of doubles                                             []
  --float64            a double arg                                                  3.12
  --i16                16-bit integer                                                1 
  --i32                32-bit integer                                                0 
  --i64                64-bit integer                                                1 
  --int64s             a list of 64-bit integers                                     []
  --is_no_bool         a boolean arg                                                 true
  --is_one_bool        a boolean arg                                                 false
  --is_rest_bool       a boolean arg                                                 true
  --is_true_bool       a boolean arg                                                 false
  --is_yes_bool        a boolean arg                                                 false
  --missing            a string arg                                                  a missing string in cfg file, default applies
  --string             a string arg                                                  A Default std::string Value
  --string.qouted      a string arg                                                  A Default std::string Value
  --strs               a list of strings                                             []
  --swc.logging.level  Logging level: debug|info|notice|warn|error|crit|alert|fatal  INFO  # (6)
Usage: %s [Options] [args]
Options
  --aGroupOne.arg.1     a group arg                                                   0 
  --aGroupOne.arg.2     a group arg                                                   0 
  --aGroupOne.arg.3     a group arg                                                   0 
  --aGroupOne.arg.4     a group arg                                                   0 
  --aGroupOne.arg.5     a group arg                                                   0 
  --app                 application filename, zero pos skipped                          
  --arg-1               argument one                                                    
  --args                rest of arguments                                             []
  --boo                 a boolean arg                                                 false
  --daemon              Start process in background mode                              true
  --debug               Shortcut to --logging-level debug                             false
  --f64s                a list of doubles                                             []
  --false               token defaults to false                                       false
  --float64             a double arg                                                  3.12
  --help -[h]           Show this help message and exit                               true
  --help-config         Show help message for config properties                       true
  --i16                 16-bit integer                                                1 
  --i32                 32-bit integer                                                0 
  --i64                 64-bit integer                                                1 
  --int64s              a list of 64-bit integers                                     []
  --is_no_bool          a boolean arg                                                 true
  --is_one_bool         a boolean arg                                                 false
  --is_rest_bool        a boolean arg                                                 true
  --is_true             if set token true                                             false
  --is_true_bool        a boolean arg                                                 false
  --is_yes_bool         a boolean arg                                                 false
  --logging-level -[l]  Logging level: debug|info|notice|warn|error|crit|alert|fatal  INFO  # (6)
  --missing             a string arg                                                  a missing string in cfg file, default applies
  --quiet               Negate verbose                                                false
  --skip1               a skipped token                                               true
  --skip2               a skipped token                                               true
  --string              a string arg                                                  A Default std::string Value
  --string.qouted       a string arg                                                  A Default std::string Value
  --strs                a list of strings                                             []
  --swc.cfg             Main configuration file                                       swc.cfg
  --swc.logging.level   Logging level: debug|info|notice|warn|error|crit|alert|fatal  INFO  # (6)
  --true                token defaults to true                                        true
  --verbose             Show more verbose output                                      false
  --version -[v]        Show version information and exit                             true
*** Raw Parsed Options:
aGroupOne.arg.1=[1]
aGroupOne.arg.2=[2]
aGroupOne.arg.3=[3]
aGroupOne.arg.4=[4]
aGroupOne.arg.5=[5]
boo=[true]
f64s=[1.123, 1.123K, 1.123M, 1.123g]
float64=[1.1234]
i16=[64000]
i32=[640K]
i64=[10G]
int64s=[1G, 2G, 3G]
is_no_bool=[false]
is_one_bool=[1]
is_rest_bool=[DefaultFalse]
is_true_bool=[True]
is_yes_bool=[yEs]
string=[aStringNoSpaces]
string.qouted=[a   S t r i n g   W i t h    S p a c e s]
strs=[StringOne, StringTwo, StringThree]
*** Parsed Options:
aGroupOne.arg.1=1
aGroupOne.arg.2=2
aGroupOne.arg.3=3
aGroupOne.arg.4=4
aGroupOne.arg.5=5
boo=true
debug=false
f64s=[1.123, 1123, 1.123e+06, 1.123e+09]
false=false
float64=1.1234
i16=64000
i32=640000
i64=10000000000
int64s=[1000000000, 2000000000, 3000000000]
is_no_bool=false
is_one_bool=true
is_rest_bool=false
is_true=false
is_true_bool=true
is_yes_bool=true
logging-level=INFO  # (6)
missing=a missing string in cfg file, default applies
quiet=false
string=aStringNoSpaces
string.qouted=a   S t r i n g   W i t h    S p a c e s
strs=[StringOne, StringTwo, StringThree]
swc.cfg=swc.cfg
swc.logging.level=INFO  # (6)
true=true
verbose=false
true
aGroupOne.arg.1=1
aGroupOne.arg.2=2
aGroupOne.arg.3=3
aGroupOne.arg.4=4
aGroupOne.arg.5=5
boo=true
debug=false (default)
dummy=true
f64s=[1.123, 1123, 1.123e+06, 1.123e+09]
false=false (default)
float64=1.1234
i16=64000
i32=640000
i64=10000000000
int64s=[1000000000, 2000000000, 3000000000]
is_no_bool=false
is_one_bool=true
is_rest_bool=false
is_true=false (default)
is_true_bool=true
is_yes_bool=true
logging-level=INFO  # (6) (default)
missing=a missing string in cfg file, default applies (default)
quiet=false (default)
string=aStringNoSpaces
string.qouted=a   S t r i n g   W i t h    S p a c e s
strs=[StringOne, StringTwo, StringThree]
swc.cfg=swc.cfg (default)
swc.logging.level=INFO  # (6) (default)
true=true (default)
verbose=false (default)
